include/openenclave/internal/sgx/td.h

```diff
@@ -0,0 +1,153 @@
+       protector.
+     */
+    uint64_t stack_guard; /* The offset is 0x28 for x64 */
+    uint64_t __reserved_0;
+    uint64_t __ssa_frame_size;
+    uint64_t __last_error;
+
+    /* The threads implementations uses this to put threads on queues */
+    oe_thread_data_t* next;
+
+    uint64_t __tls_addr;
+    uint64_t __tls_array;
+    uint64_t __exception_flag; /* number of exceptions being handled */
+    uint64_t __cxx_thread_info[6];
+
+    // The exception code.
+    uint32_t exception_code;
+    // The exception flags.
+    uint32_t exception_flags;
+    // The rip when exception happened.
+    uint64_t exception_address;
+};
+
+OE_CHECK_SIZE(sizeof(oe_thread_data_t), 168);
+
+oe_thread_data_t* oe_get_thread_data(void);
+
+/*
+**==============================================================================
+**
+** td_t
+**
+**     Extended thread data
+**
+**==============================================================================
+*/
+
+#define TD_MAGIC 0xc90afe906c5d19a3
+
+#define OE_THREAD_LOCAL_SPACE (OE_PAGE_SIZE)
+
+/**
+ * thread_specific_data is the last field in td_t. It takes up any remainaing
+ * space after the declarations of the previous fields.
+ * Its size is equal to
+ *     sizeof(td_t) - OE_OFFSETOF(td_t, thread_specific_data)
+ * Due to the inability to use OE_OFFSETOF on a struct while defining its
+ * members, this value is computed and hard-coded.
+ */
+#define OE_THREAD_SPECIFIC_DATA_SIZE (3824)
+
+typedef struct _callsite Callsite;
+
+/* Thread specific TLS atexit call parameters */
+typedef struct _oe_tls_atexit
+{
+    void (*destructor)(void*);
+    void* object;
+} oe_tls_atexit_t;
+
+OE_PACK_BEGIN
+typedef struct _td
+{
+    oe_thread_data_t base;
+
+    /* A "magic number" for sanity checking (TD_MAGIC) */
+    uint64_t magic;
+
+    /* Depth of ECALL stack (zero indicates that it is unwound) */
+    uint64_t depth;
+
+    /* Host registers saved here on entry and restored on exit */
+    uint64_t host_rcx; /* EENTER return address */
+    uint64_t host_rsp;
+    uint64_t host_rbp;
+    uint64_t host_previous_rsp;
+    uint64_t host_previous_rbp;
+
+    /* Return arguments from OCALL */
+    uint16_t oret_func;
+    uint16_t oret_result;
+    uint16_t padding[2];
+    uint64_t oret_arg;
+
+    /* List of Callsite structures (most recent call is first) */
+    Callsite* callsites;
+
+    /* Simulation mode is active if non-zero */
+    uint64_t simulate;
+
+    /* Host ecall context pointers */
+    struct _oe_ecall_context* host_ecall_context;
+    struct _oe_ecall_context* host_previous_ecall_context;
+
+    /* Reserved for thread specific data. */
+    uint8_t thread_specific_data[OE_THREAD_SPECIFIC_DATA_SIZE];
+} td_t;
+OE_PACK_END
+
+OE_CHECK_SIZE(sizeof(td_t), 4096);
+OE_STATIC_ASSERT(
+    OE_THREAD_SPECIFIC_DATA_SIZE ==
+    sizeof(td_t) - OE_OFFSETOF(td_t, thread_specific_data));
+
+/* Get the thread data object for the current thread */
+td_t* oe_get_td(void);
```

Comment:
Question: the API oe_get_id() doesn't have sgx in the name.  Since this is under sgx/td.h do we think other TEEs in the future would need TD?  If so, should td_t be called sgx_td_t and oe_get_td be renaed to oe_get_sgx_td to avoid naming conflicts?

Reply:
td_t as it stands today is at least x86_64 specific (it saves rsp, rbp, etc.). @anakrish do you see any issue with changing the naming to be sgx-specific?

Reply:
I don't see any issue with renaming the function. 

However,  I do think that having platform specific prefixes first needs to be made a development guideline if it is already not so.

Reply:
Good point. I don't really know the answer so I opened https://github.com/openenclave/openenclave/issues/2796 to discuss